### PR TITLE
Accurately set the lft and rght fields for moveDown

### PR DESCRIPTION
**Pull request by AD7six**
Needs similar updates to moveUp, fixes https://github.com/cakephp/cakephp/issues/5863.

Will cause some test fails because.. it looks like [this method didn't say what it was doing - but instead reported something different](https://github.com/cakephp/cakephp/compare/3.0...3.0-tree-wrong).

### CODE REVIEW

**Comment by markstory** [comment]
Seems reasonable to me.

**Comment by ADmad** [comment]
Tests failing.

**Comment by AD7six** [comment]
> Tests failing.

But not the added ones :trollface: 

I need to pick the test case apart to determine which fails need fixing, and which passes are actually broken (if any).

**Comment by ADmad** [comment]
For one failing test the actual value for lft is negative so something is fishy :stuck_out_tongue:

**Comment by AD7six** [comment]
The original problem this PR fixed was that moveDown - would return a node with lft/rght set to different values than were on the database, and the test case would only be asserting the nodes (transient) state whilst leaving the db in a _different_ state (saying it moved a node 2 places, and instead moving it 1).

This PR has morphed a bit to include:

 * Updating the test case to be easier to read/understand/maintain.
 * Update moveUp and moveDown to be easier to read/understand/maintain.
 * Updating moveUp and moveDown to move in one step (3 update queries) instead of once per move ( 3 * $number queries)

@lorenzo Because I've changed the implementation a fair bit - can you give it a look over please.

File: src/ORM/Behavior/TreeBehavior.php [hunk 1]
```diff
@@ -557,51 +557,58 @@ public function moveUp(Entity $node, $number = 1)
...
-        list($parent, $left, $right) = [$config['parent'], $config['left'], $config['right']];
-
         if (!$number) {
             return false;
         }
-
-        $parentLeft = 0;
-        if ($node->get($parent)) {
-            $parentLeft = $this->_getNode($node->get($parent))->get($left);
+        if ($number < 0) {
```
Comment (AD7six) [inline src/ORM/Behavior/TreeBehavior.php]: @lorenzo this is one of the things I wanted specifically to check - it still returns the node, [as it did before](https://github.com/cakephp/cakephp/blob/3.0/tests/TestCase/ORM/Behavior/TreeBehaviorTest.php#L290) - is that desired behavior?

Comment (markstory) [inline src/ORM/Behavior/TreeBehavior.php]: I think the intent would have been to catch negatives as well. Negative values make no sense in MPTT.

File: tests/TestCase/ORM/Behavior/TreeBehaviorTest.php [hunk 2]
```diff
@@ -284,17 +339,49 @@ public function testMoveUp()
...
         $this->assertEquals(['lft' => 1, 'rght' => 10], $node->extract(['lft', 'rght']));
+        $expected = [
+            ' 1:10 -  1:Link 1',
+            '_ 2: 3 -  2:Link 2',
+            '_ 4: 9 -  3:Link 3',
+            '__ 5: 8 -  4:Link 4',
+            '___ 6: 7 -  5:Link 5',
+            '11:14 -  6:Link 6',
+            '_12:13 -  7:Link 7',
+            '15:16 -  8:Link 8'
```
Comment (markstory) [inline tests/TestCase/ORM/Behavior/TreeBehaviorTest.php]: I really like this style for the tests. Much simpler to read.

File: src/ORM/Behavior/TreeBehavior.php (hunk 1 above)
Comment (AD7six) [inline src/ORM/Behavior/TreeBehavior.php]: I agree, though the same applies whatever the implementation (move this foo up -1 places).

I'll change it to treat < 1 as invalid.

**Comment by AD7six** [comment]
I think this is good enough to merge, if it breaks something it should hopefully be easier now to write a test for whatever scenario and fix it.

**Pull request merged**